base/system/winlogon/rpcserver.c

```diff
@@ -499,20 +499,102 @@ BaseInitiateSystemShutdownEx(
     BOOLEAN bRebootAfterShutdown,
     ULONG dwReason)
 {
+    HANDLE hToken;
+    PTOKEN_PRIVILEGES tpPrivileges;
+    LUID luidSysShutdown;
+    LUID luidRemoteSysShutdown;
+    LPVOID lpAclBuffer;
+    DWORD dwAclSize = 0;
+    
     TRACE("BaseInitiateSystemShutdownEx()\n");
     TRACE("  Message: %wZ\n", lpMessage);
     TRACE("  Timeout: %lu\n", dwTimeout);
     TRACE("  Force: %d\n", bForceAppsClosed);
     TRACE("  Reboot: %d\n", bRebootAfterShutdown);
     TRACE("  Reason: %lu\n", dwReason);
 
-    //FIXME: Verify that the caller actually has the correct privileges
+    //FIXME: Verify that the caller actually has the correct privileges for remote shutdown
+  // Check if thread token exists
+    if (!OpenThreadToken(GetCurrentThread(), TOKEN_ALL_ACCESS, FALSE, &hToken))
```

Comment:
You're getting the current process's (winlogon's) token here, however we really need to check the RPC caller's token. That means you should call RpcImpersonateClient first, and then RpcRevertToSelf on exit.